price-feeder/oracle/provider/kraken.go

```diff
@@ -23,100 +26,385 @@ type (
 	if err != nil {
-		return nil, fmt.Errorf("failed to read Kraken response body: %w", err)
+		p.logger.Err(err).Msg("Kraken provider could not marshal ticker message")
+		return
 	}
 
-	var tickerResp KrakenTickerResponse
-	if err := json.Unmarshal(bz, &tickerResp); err != nil {
-		return nil, fmt.Errorf("failed to unmarshal Kraken response body: %w", err)
+	var krakenTicker KrakenTicker
+	if err := json.Unmarshal(tickerBz, &krakenTicker); err != nil {
+		p.logger.Err(err).Msg("Kraken provider could not unmarshal ticker")
+		return
 	}
 
-	if len(tickerResp.Error) != 0 {
-		return nil, fmt.Errorf("received unexpected error from Kraken response: %v", tickerResp.Error)
+	krakenPair, ok := tickerMessage[3].(string)
+	if !ok {
+		p.logger.Debug().Msg("Kraken provider sent an unexpected pair")
+		return
 	}
 
-	if len(tickers) != len(tickerResp.Result) {
-		return nil, fmt.Errorf(
-			"received unexpected number of tickers; expected: %d, got: %d",
-			len(tickers), len(tickerResp.Result),
-		)
+	krakenPair = normalizeKrakenPair(krakenPair)
+	currencyPairSymbol := krakenPairToCurrencyPairSymbol(krakenPair)
+
+	tickerPrice, err := krakenTicker.toTickerPrice(currencyPairSymbol)
+	if err != nil {
+		p.logger.Err(err).Msg("Kraken provider could not parse kraken ticker to ticker price")
+		return
 	}
 
-	tickerPrices := make(map[string]TickerPrice, len(tickers))
-	for _, t := range tickers {
-		// TODO: We may need to transform 't' prior to looking it up in the response
-		// as Kraken may represent currencies differently.
-		pair, ok := tickerResp.Result[t]
-		if !ok {
-			return nil, fmt.Errorf("failed to find ticker in Kraken response: %s", t)
-		}
+	p.setTickerPair(currencyPairSymbol, tickerPrice)
+}
 
-		price, err := sdk.NewDecFromStr(pair.C[0])
-		if err != nil {
-			return nil, fmt.Errorf("failed to parse Kraken price (%s) for %s", pair.C[0], t)
-		}
+// reconnect closes the last WS connection and create a new one.
+func (p *KrakenProvider) reconnect() error {
+	p.wsClient.Close()
+
+	wsConn, _, err := websocket.DefaultDialer.Dial(p.wsURL.String(), nil)
+	if err != nil {
+		return fmt.Errorf("error connecting to Kraken websocket: %w", err)
+	}
+	p.wsClient = wsConn
+
+	pairs := make([]string, len(p.subscribedPairs))
+	iterator := 0
+	for _, cp := range p.subscribedPairs {
+		pairs[iterator] = currencyPairToKrakenPair(cp)
+		iterator++
+	}
 
-		volume, err := sdk.NewDecFromStr(pair.V[1])
-		if err != nil {
-			return nil, fmt.Errorf("failed to parse Kraken volume (%s) for %s", pair.V[1], t)
+	return p.subscribePairs(pairs...)
+}
+
+// keepReconnecting keeps trying to reconnect if an error occurs in recconnect.
+func (p *KrakenProvider) keepReconnecting() {
+	reconnectTicker := time.NewTicker(defaultReconnectTime)
+	defer reconnectTicker.Stop()
+
+	for time := range reconnectTicker.C {
+		if err := p.reconnect(); err != nil {
+			p.logger.Err(err).Msg("Kraken provider received recconecting at " + time.String())
```

Comment:
nit:
```suggestion
			p.logger.Err(err).Msg("kraken provider attempted reconnecting at " + time.String())
```
Also, what happens if the kraken provider never reconnects?

Reply:
basically, it would keep trying to reconnect at each 15m